### PR TITLE
Implement `vec_chop(sizes = )`

### PR DESCRIPTION
**Pull request by DavisVaughan**
Closes https://github.com/r-lib/vctrs/issues/1598
Closes https://github.com/r-lib/vctrs/issues/1210

This PR implements a new `sizes` argument to `vec_chop()`, with the following specs:
- `sizes` must be an integer vector containing only non-negative sizes (`0` is ok, but `-1` and `NA` are not)
- `sizes` must sum to `vec_size(x)`
- `sizes` can't be used alongside `indices`

`sizes` allows us to efficiently chop a vector using the results of either `vec_run_sizes()` or `list_sizes()`. It is typically faster than chopping with locations from `vec_group_loc()` if you know the input is ordered (or already grouped).

```r
set.seed(123)

x <- sample(1e5, 1e6, replace = TRUE)
x <- sort(x)

head(x, n = 20)
#>  [1] 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 3

# Split `x` on its groups
bench::mark(
  sizes = vec_chop(x, sizes = vec_run_sizes(x)),
  indices = vec_chop(x, indices = vec_group_loc(x)$loc)
)
#> # A tibble: 2 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 sizes        8.37ms   10.7ms      94.9     2.1MB     9.49
#> 2 indices     52.61ms   53.1ms      18.9    20.2MB    31.4
```

With ^, typically you chop something else, but you use the run sizes of `x` to index it.

It also allows for some neat patterns with `list_sizes()` related to "unchop, do something, rechop" that were clunky before:

``` r
x <- list(c(1, 2, 1), c(3, 1), 5, double())

x_flat <- list_unchop(x)
x_flat <- x_flat + max(x_flat)

vec_chop(x_flat, sizes = list_sizes(x))
#> [[1]]
#> [1] 6 7 6
#> 
#> [[2]]
#> [1] 8 6
#> 
#> [[3]]
#> [1] 10
#> 
#> [[4]]
#> numeric(0)
```

---

I've spent a significant amount of time cleaning up the `vec_chop()` internals. I think they are now much easier to read. In particular, `vctrs_chop_info` was doing way too much:
- `vctrs_chop_info` was managing the `out` container, which it no longer does
- `vctrs_chop_info` was taking and managing the `proxy` info, which it no longer does

I've now narrowly scoped `vctrs_chop_info` to only manage the index generation process, and have renamed it to `vctrs_chop_indices`. It handles 3 different paths depending on which of the following cases we are in:
- `indices = NULL, sizes = NULL`
- `indices != NULL`
- `sizes != NULL`

That seemed to dramatically clean things up. I think if you just skim over the new `slice-chop.c` without the diffs, then that should be apparent.

### CODE REVIEW

File: R/slice-chop.R [hunk 1]
```diff
@@ -89,8 +105,25 @@
...
+#'
+#' # If you have a list of homogeneous vectors, sometimes it can be useful to
+#' # unchop, apply a function to the flattened vector, and then rechop according
+#' # to the original indices. This can be done efficiently with `list_sizes()`.
+#' x <- list(c(1, 2, 1), c(3, 1), 5, double())
+#' x_flat <- list_unchop(x)
+#' x_flat <- x_flat + max(x_flat)
+#' vec_chop(x_flat, sizes = list_sizes(x))
+vec_chop <- function(x, indices = NULL, ..., sizes = NULL) {
+  check_dots_empty0(...)
```
Comment (DavisVaughan) [inline R/slice-chop.R:125]: I don't think we can move `indices` at this point, but I wanted to add the empty dots in and felt it was worth it to do so anyways

File: src/bind.c [hunk 2]
```diff
@@ -304,7 +304,7 @@ r_obj* as_df_row_impl(r_obj* x,
   // Remove names first as they are promoted to data frame column names
   x = KEEP(vec_set_names(x, r_null));
 
-  x = KEEP(vec_chop(x, r_null));
+  x = KEEP(vec_chop_unsafe(x, r_null, r_null));
```
Comment (DavisVaughan) [inline src/bind.c:307]: The version of `vec_chop()` we use internally has been renamed to `vec_chop_unsafe()` to make it clear that it doesn't perform `vec_as_location()` or `vec_as_chop_sizes()` checks on the inputs.

Similar to `vec_slice_unsafe()`!

File: src/c-unchop.c [hunk 3]
```diff
@@ -85,7 +85,7 @@ r_obj* list_unchop(r_obj* xs,
     out_size += r_length(r_list_get(indices, i));
   }
 
-  r_obj* locs = KEEP(vec_as_indices(indices, out_size, r_null));
+  r_obj* locs = KEEP(list_as_locations(indices, out_size, r_null));
```
Comment (DavisVaughan) [inline src/c-unchop.c:88]: `vec_as_indices()` took a list of location vectors and called `vec_as_location()` on each. I've renamed that to `list_as_locations()` to reflect modern naming schemes. It might not be a perfect name but I think it is much closer.

File: src/callables.c [hunk 4]
```diff
@@ -19,7 +19,7 @@ SEXP exp_vec_cast(SEXP x, SEXP to) {
 }
 
 SEXP exp_vec_chop(SEXP x, SEXP indices) {
-  return vec_chop(x, indices);
+  return vec_chop_unsafe(x, indices, r_null);
 }
```
Comment (DavisVaughan) [inline src/callables.c:23]: Exposed experimental `exp_vec_chop()` callable does not gain the `sizes` argument

File: tests/testthat/_snaps/c.md [hunk 5]
```diff
@@ -256,13 +256,13 @@
...
     Code
       with_memory_prof(list_unchop(make_list_of(2000)))
     Output
-      [1] 220KB
+      [1] 205KB
     Code
       with_memory_prof(list_unchop(make_list_of(4000)))
     Output
-      [1] 439KB
+      [1] 408KB
```
Comment (DavisVaughan) [inline tests/testthat/_snaps/c.md:267]: `make_list_of()` calls `vec_chop()` so I'm not entirely surprised to see changes here, but I don't really have a clue what the changes are attributed to

File: tests/testthat/_snaps/slice-chop.md [hunk 6]
```diff
@@ -1,3 +1,82 @@
...
+---
+
+    Code
+      (expect_error(vec_chop(1, indices = list(1.5)), class = "vctrs_error_subscript_type")
+      )
+    Output
+      <error/vctrs_error_subscript_type>
+      Error:
+      ! Can't subset elements.
+      x Can't convert from <double> to <integer> due to loss of precision.
```
Comment (DavisVaughan) [inline tests/testthat/_snaps/slice-chop.md:18]: The errors of `vec_chop()` aren't perfect, but that is partially due to not having exposed `error_call` and `error_arg` like `list_unchop()`. But that is big enough to be its own PR. I have moved a few tests towards snapshot tests rather than just `expect_error()` calls so we will eventually see the improvements when we add those arguments.

File: src/slice-chop.c [hunk 7]
```diff
@@ -2,60 +2,144 @@
...
 
 static
-struct vctrs_chop_info new_chop_info(r_obj* x, r_obj* indices) {
-  struct vctrs_chop_info info;
-  info.shelter = KEEP(r_alloc_list(3));
+struct vctrs_chop_indices* new_chop_indices(r_obj* x, r_obj* indices, r_obj* sizes) {
+  r_obj* shelter = KEEP(r_alloc_list(4));
+
+  r_obj* self = r_alloc_raw(sizeof(struct vctrs_chop_indices));
+  r_list_poke(shelter, 0, self);
```
Comment (DavisVaughan) [inline src/slice-chop.c:70]: Also switched to our more standard approach of returning a pointer that manages itself

File: src/slice-chop.h [hunk 8]
```diff
@@ -3,8 +3,10 @@
 
 #include "vctrs-core.h"
 
+r_obj* vec_chop(r_obj* x, r_obj* indices, r_obj* sizes);
+r_obj* vec_chop_unsafe(r_obj*, r_obj* indices, r_obj* sizes);
```
Comment (DavisVaughan) [inline src/slice-chop.h:7]: Moved `vec_chop()` from `<vctrs.h>` to `<slice-chop.h>`

And of course now `vec_chop()` is the version that does the index checks, and `vec_chop_unsafe()` is the performance variant that assumes the indices are already correct

**Review by lionel- — APPROVED** [review]
Looks great!

File: R/slice-chop.R (hunk 1 above)
Comment (DavisVaughan) [reply R/slice-chop.R]: Good idea, done here https://github.com/r-lib/vctrs/pull/1813/commits/c1bd65e261efcba34b7813d5e4f38abd02287c85
